Automattic/jetpack
# Featured images fallback: add srcset, width/height, and loading attributes

In an effort to move away from the `devicepx` script towards native functionality, this PR adds a `srcset` to qualifying featured images obtained via `jetpack_featured_images_fallback_get_image` (which is used as a filter on `post_thumbnail_html`). The browser will automatically pick the most appropriate image from the options in the `srcset`, depending on the screen pixel ratio and the current zoom value.

Qualifying images also get `width` and `height` to avoid layout shifts.

Furthermore, all images get the `loading` attribute explicitly defined for an ancillary performance improvement: `lazy` on non-singular pages where several featured images will most likely be loaded, and `eager` on singular pages where the featured image will likely be at the top and is thus a strong candidate for the Largest Contentful Paint element.

Qualifying images are defined as those for which we know the source width and height at serving time, which typically means they're attachment images.

**Important note regarding wpcom:** as far as I can tell, this change will have no effect on wpcom, since a different codepath is used there instead, in a nearly-identical file. I will create a companion diff to this PR in wpcom once these changes are accepted.

## Proposed changes:
* Add support for width-based `srcset`s to `Jetpack_PostImages::generate_cropped_srcset`
* For qualifying images, add `srcset` to the image returned by `jetpack_featured_images_fallback_get_image`
* For qualifying images, add `width` and `height` to the image returned by `jetpack_featured_images_fallback_get_image`
* Add `loading=lazy` by default to all images returned by `jetpack_featured_images_fallback_get_image` on non-singular pages, and `loading=eager` by default to all images returned by `jetpack_featured_images_fallback_get_image` on singular pages

### Other information:

- N/A Have you written new tests for your changes, if applicable?

## Jetpack product discussion
None.

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:

Start by creating at least two blog posts, with one image each:
- One blog post with a large attachment image, i.e., an image uploaded to the site
- One blog post with a large external image, i.e., an image provided from an external URL

For themes that use cropped thumbnails:
- Apply a theme with cropped thumbnails, such as Gazette
- Visit a blog post with an external image
  - Inspect the featured image (in the case of Gazette, it's a banner at the top)
  - Ensure the featured image is correctly resized and cropped to the theme's desired aspect ratio
  - Ensure the featured image has the `loading` attribute set to `eager`
- Visit a blog post with an attachment image
  - Inspect the featured image (in the case of Gazette, it's a banner at the top)
  - Ensure the featured image is correctly resized and cropped to the theme's desired aspect ratio
  - Ensure the featured image has the `loading` attribute set to `eager`
  - Ensure the featured image has a `srcset` attribute with multiple sources defined
  - Ensure the featured image has defined `width` and `height` attributes
- Visit a directory page, e.g. `/2023/`
  - Inspect the featured images (in the case of Gazette, they're only the post thumbnails defined as `<img>` tags with the `wp-post-image` class; some other images are style backgrounds instead and are not relevant here)
  - Ensure they have the `loading` attribute set to `lazy`

For themes that use uncropped thumbnails:
- Apply a theme with uncropped thumbnails, such as Baskerville 2
- Visit a blog post with an external image
  - Inspect the featured image (in the case of Baskerville 2, it's the large image at the top)
  - Ensure the featured image is correctly resized and uncropped, maintaining the original aspect ratio
  - Ensure the featured image has the `loading` attribute set to `eager`
- Visit a blog post with an attachment image
  - Inspect the featured image (in the case of Baskerville 2, it's the large image at the top)
  - Ensure the featured image is correctly resized and uncropped, maintaining the original aspect ratio
  - Ensure the featured image has the `loading` attribute set to `eager`
  - Ensure the featured image has a `srcset` attribute with multiple sources defined
  - Ensure the featured image has defined `width` and `height` attributes
- Visit a directory page, e.g. `/2023/`
  - Inspect the featured images (in the case of Baskerville 2, they're the post thumbnails, and they're inside `div`s with the `featured-media` class)
  - Ensure they have the `loading` attribute set to `lazy`
